app/ts/simulation/priceEstimator.ts

```diff
@@ -21,99 +20,92 @@ export class PriceEstimator {
-				inOutResults.push( {
-					token: addressString(token.token ),
-					inOutAmount: [inOut.amounts[1], inOut.amounts[0]],
-					decimals: token.decimals
-				})
+			const poolAddresses = calculateUniswapPools(token.address, CHAINS[chainString].weth)
+			const uniswapSpotCalls = getUniswapSpotCalls(token.address, quoteToken.address, poolAddresses)
+
+			const callData = stringToUint8Array(IMulticall3.encodeFunctionData('aggregate3', [uniswapSpotCalls]))
+			const callTransaction = {
+				type: '1559',
+				to: MULTICALL3,
+				value: 0n,
+				input: callData,
 			}
+			const multicallReturnData: { success: boolean, returnData: string }[] = IMulticall3.decodeFunctionResult('aggregate3', await this.ethereum.call(callTransaction, 'latest'))[0]
+
+			const prices = multicallReturnData.reduce((prices, { success, returnData }, index): { price: bigint, k: bigint }[] => {
+				if (index > 4) return prices
+
+				// V2 Pool
+				if (index === 0 && success) {
+					const { reserve0, reserve1 } = IUniswapV2Pool.decodeFunctionResult('getReserves', returnData)
+
+					const price = poolAddresses.token0IsQuote
+						? (reserve1 * (10n ** BigInt(token.decimals))) / reserve0
+						: (reserve0 * (10n ** BigInt(token.decimals))) / reserve1
+
+					return [...prices, { price, k: BigInt(reserve0 * reserve1) }]
+				}
+
+				// V3 Pool
+				if (success && multicallReturnData[index * 2 + 3].success && multicallReturnData[index * 2 + 4].success) {
+					const { sqrtPriceX96 } = IUniswapV3Pool.decodeFunctionResult('slot0', returnData)
+
+					const price = poolAddresses.token0IsQuote
+						? ((sqrtPriceX96 * (10n ** BigInt(token.decimals)) / (2n ** 96n)) ** 2n) / (10n ** BigInt(token.decimals))
+						: (10n ** (18n)) / ((sqrtPriceX96 / (2n ** 96n)) ** 2n)
+
+					const reserve0 = BigInt(multicallReturnData[index * 2 + 3].returnData)
+					const reserve1 = BigInt(multicallReturnData[index * 2 + 4].returnData)
+
+					return [...prices, { price, k: reserve0 * reserve1 }]
+				}
+
+				return prices
+			}, [] as { price: bigint, k: bigint }[])
+
+			// Use pool with most TVL
+			const price = prices.length > 0 ? prices.reduce((highestLiq, p) => p.k > highestLiq.k ? p : highestLiq).price : undefined
+
+			if (price) tokenPrices.push({
+				token,
+				quoteToken,
+				price
+			})
 		}
-		return inOutResults
+		return tokenPrices
 	}
 }
 
-export function getTokenPrice(ethInOutAmounts: readonly bigint[], tokenDecimals: bigint, precision: number) {
-	return Number(ethInOutAmounts[1] * 10n ** BigInt(precision) / ethInOutAmounts[0]) / Number(10n ** BigInt(precision)* 10n ** tokenDecimals / 10n ** 18n)
-}
-
 export function getTokenAmountsWorth(tokenAmount: bigint, tokenPriceEstimate: TokenPriceEstimate) {
-	return (tokenPriceEstimate.inOutAmount[0] * tokenAmount / tokenPriceEstimate.inOutAmount[1])
+	return (tokenPriceEstimate.price * tokenAmount) / (10n ** tokenPriceEstimate.token.decimals)
```

Comment:
I feel this is wrong if the price of a token is small enough? as the price would just be 0